Orleans.Providers.MongoDB/Reminders/Store/MongoReminderCollection.cs

```diff
@@ -129,72 +152,59 @@ public async Task<bool> RemoveRow(GrainId grainId, string reminderName, string e
-                    await Collection.FindOneAndUpdateAsync<MongoReminderDocument, MongoReminderDocument>(x => x.Id == id && x.Etag == eTag,
-                        Update.Set(x => x.IsDeleted, true),
-                        FindAndUpsert);
-
-                await Collection.DeleteManyAsync(x => x.IsDeleted);
-
-                return string.Equals(existingDocument?.ReminderName, reminderName, StringComparison.Ordinal);
-            }
-            catch (MongoException ex)
-            {
-                if (ex.IsDuplicateKey())
-                {
-                    return false;
-                }
-                throw;
-            }
+            var deleteResult = await Collection.DeleteOneAsync(x => x.Id == id && x.Etag == eTag);
+            return deleteResult.DeletedCount > 0;
         }
 
         public virtual Task RemoveRows()
         {
-            return Collection.DeleteManyAsync(r =>
-                // note: the query is to align to the indexes (ByHash being the primary guideline)
-                (r.IsDeleted == false && r.ServiceId == serviceId) ||
-                (r.IsDeleted == true && r.ServiceId == serviceId)
-            );
+            // note: only used and called by the test harness
+            return Collection.DeleteManyAsync(r => r.ServiceId == serviceId);
         }
 
         public virtual async Task<string> UpsertRow(ReminderEntry entry)
         {
             var id = ReturnId(serviceId, entry.GrainId, entry.ReminderName);
+            var document = MongoReminderDocument.Create(id, serviceId, entry, Guid.NewGuid().ToString());
 
-            var updatedEtag = Guid.NewGuid().ToString();
-            var updateDocument = MongoReminderDocument.Create(id, serviceId, entry, updatedEtag);
+            var useUpsert = entry.ETag != null || !await TryInsertOneAsync();
 
-            try
+            if (useUpsert)
             {
-                await Collection.ReplaceOneAsync(x => x.Id == id,
-                    updateDocument,
-                    UpsertReplace);
+                // see comments in TryInsertOneAsync to determine when selecting the upsert.
+                await Collection.ReplaceOneAsync(x => x.Id == id, document, UpsertReplace);
             }
-            catch (MongoException ex)
+
+            return entry.ETag = document.Etag;
+
+            async Task<bool> TryInsertOneAsync()
             {
-                if (!ex.IsDuplicateKey())
+                // when the etag is null, it is a strong indicator that an insertion is only necessary
+                // as it is brand new.
+                // In the unlikely event that this assumption is incorrect, mongo will throw a conflict.
+                try
+                {
+                    // insertion is a lot faster than doing a search in Mongo
+                    await Collection.InsertOneAsync(document);
+                    return true;
+                }
+                catch (MongoCommandException ex)
```

Comment:
This is terminating the silo due to unhandled exceptions in scenarios where mongo driver is implicitly batching concurrent operations, please update the handled exception type to `MongoException`

You can see the issue by updating `TestReminders` method **TestHost** project to the following:
```
        private static async Task TestReminders(IClusterClient client)
        {
            await Task.WhenAll(Enumerable.Range(1, 5).Select(async i =>
            {
                var reminderGrain = client.GetGrain<IReminderGrain>(1);

                await reminderGrain.StartReminder("TestReminder", TimeSpan.FromMinutes(1));
            }));
        }
```